### PR TITLE
Simplify unions when erasing last known values

### PR DESCRIPTION
**Pull request by JukkaL**
When we erase last known values in an union with multiple
Instance types, make sure that the resulting union doesn't have
duplicate erased types. The duplicate items weren't incorrect as such,
but they could cause overly complex error messages and potentially
slow type checking performance.

This is one of the fixes extracted from #12054. Since some of the
changes may cause regressions, it's better to split the PR.

Work on #12051.

### CODE REVIEW

**Comment by JukkaL** [comment]
This seems to have some unintended consequences. I'm going to try making the union simplification less eager.

**Comment by JukkaL** [comment]
@sobolevn This is similar to the first half of #12054, but this only simplifies in the cases that we care about to avoid changes in external behavior. Using `make_simplified_union` always was a too big and unnecessary change. It would be great if you can have another look.

File: mypy/test/typefixture.py [hunk 1]
```diff
@@ -157,9 +157,11 @@ def make_type_var(name: str, id: int, values: List[Type], upper_bound: Type,
         self.lit1 = LiteralType(1, self.a)
         self.lit2 = LiteralType(2, self.a)
         self.lit3 = LiteralType("foo", self.d)
+        self.lit4 = LiteralType(3, self.a)
```
Comment (sobolevn) [inline mypy/test/typefixture.py:160]: nit:

```suggestion
        self.lit4 = LiteralType(4, self.a)
```

**Review by sobolevn — APPROVED** [review]
Thanks, I start to understand what is going on with this bug 🙂

**Pull request merged**